### DISCUSSION TITLE
How to deploy webpack SPA to Netlify

### DISCUSSION THREAD

**Discussion by GonzalesMatthew** [nss-evening-cohort-14/evening-client-side#132] [Help Me!] [answered]
Matthew Gonzales

Have you searched the other help tickets?
Yes. No similar issues

Describe your issue
Failing to deploy finished Bear Watcher App to Netlify

List any error messages that appear
Deploy did not succeed: Deploy directory 'dist' does not exist

What have you googled? (Paste links to the things you have found from your google search - You must include at least 3 links)
https://docs.netlify.com/configure-builds/common-configurations/
https://docs.netlify.com/configure-builds/get-started/
https://stackoverflow.com/questions/62122571/how-do-deploy-a-site-on-netlify-thats-build-with-webpack-commands

What have you tried? Tell us everything you have tried.
I read the first link which documents what to do, but still not clear to me. Do I touch a dist folder in my project? But what do I put in it? Also seeing in .gitignore of project that dist/ is included in there. 

Copy/paste link TO THE FILE on your branch or in your repo that you are having issues with NO SCREEN SHOTS 🥰
NOTE: We will not accept links to the whole repo. You must provide a link directly to the specific file
N/A - my repo is fine and runs ESLint and console error-free. Just trying to understand how to deploy correctly.

**Comment by drteresavasquez** [comment, marked as answer]
I don't think we talked deeply on deployment for webpack and Firebase specifically but READ the webpack template. https://github.com/nss-evening-cohort-14/webpack-template

There is a section on deployment that we reviewed in class the first time we used it. ALWAYS READ READ READ. 

Here also is another walk through regarding deployment. Start at the Let's Deploy section. https://github.com/drteresavasquez/deploy-react-app-with-netlify#lets-deploy

**Comment by cmeffley** [comment]
I am having the same issue. There is no `.env` file for this project and I know it's not required since there is no API keys. So I feel I am missing something else, but do not know what.

**Reply by drteresavasquez** [reply]
Did you follow the deploy steps on the readme? Also read the logs in the deploy.

**Reply by cmeffley** [reply]
Fixed it. I apparently skipped over the publish directory, thinking it auto-populated the `build`.  I again missed another little thing, so I now have a sticky note next to my computer that says "Look at EVERYTHING".

**Reply by drteresavasquez** [reply]
🤣 no worries. Know that we are trying to make sure you all have all the info you need so the readme will contain most info. I'm glad you found it!

**Comment by danicrosby** [comment]
@TrinityChristiana @ajawashington @drteresavasquez 

I am trying to update the Netlify Deployment with no success. I have read down through the discussion tickets and have done what the NPM Build and Build commands and still fails to deploy. I removed the directions off my Readme, so I'm now not sure what I'm missing to get this to work. I have a json.lock in my file structure that looks new- 


![Screen Shot 2021-02-27 at 12 36 12 PM](https://user-images.githubusercontent.com/68397076/109396682-62d0b680-78f8-11eb-86cf-0b6b41bd3afb.png)

![Screen Shot 2021-02-27 at 12 34 48 PM](https://user-images.githubusercontent.com/68397076/109396650-3157eb00-78f8-11eb-83bb-4ad8291ef4fb.png)

![Screen Shot 2021-02-27 at 12 33 45 PM](https://user-images.githubusercontent.com/68397076/109396617-0cfc0e80-78f8-11eb-9461-c6c679d76bb1.png)

**Reply by TrinityChristiana** [reply]
@danicrosby Can you send a screenshot of your last deployment console in Netlify?

**Reply by TrinityChristiana** [reply]
So if you go to your project in Netlify, click on Deploy, select the last failed deploy, and send a screenshot of the Deploy Log

### Example of a Log
![Screen Shot 2021-02-27 at 1 03 20 PM](https://user-images.githubusercontent.com/31781724/109397441-2dc66300-78fc-11eb-9976-5f1c8fb17943.png)

**Reply by danicrosby** [reply]
8:07:29 PM: Build ready to start
8:07:30 PM: build-image version: d84c79427e8f83c1ba17bcdd7b3fe38059376b68
8:07:30 PM: build-image tag: v3.6.1
8:07:30 PM: buildbot version: 4c3d8713ccd4c98ab9180f79a013d42d2076d9ff
8:07:31 PM: Fetching cached dependencies
8:07:31 PM: Failed to fetch cache, continuing with build
8:07:31 PM: Starting to prepare the repo for build
8:07:31 PM: No cached dependencies found. Cloning fresh repo
8:07:31 PM: git clone https://github.com/danicrosby/bear-watcher
8:07:32 PM: Preparing Git Reference refs/heads/main
8:07:33 PM: Different publish path detected, going to use the one specified in the Netlify configuration file: 'build' versus 'build/' in the Netlify UI
8:07:33 PM: Starting build script
8:07:33 PM: Installing dependencies
8:07:33 PM: Python version set to 2.7
8:07:35 PM: v12.18.0 is already installed.
8:07:35 PM: Now using node v12.18.0 (npm v6.14.4)
8:07:35 PM: Started restoring cached build plugins
8:07:35 PM: Finished restoring cached build plugins
8:07:35 PM: Attempting ruby version 2.7.1, read from environment
8:07:37 PM: Using ruby version 2.7.1
8:07:37 PM: Using PHP version 5.6
8:07:37 PM: Started restoring cached node modules
8:07:37 PM: Finished restoring cached node modules
8:07:37 PM: Installing NPM modules using NPM version 6.14.4
8:07:39 PM: npm WARN deprecated eslint-loader@4.0.2: This loader has been deprecated. Please use eslint-webpack-plugin
8:07:39 PM: npm WARN deprecated axios@0.20.0: Critical security vulnerability fixed in v0.21.1. For more information, see https://github.com/axios/axios/pull/3410
8:07:40 PM: npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
8:07:41 PM: npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
8:07:42 PM: npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
8:07:45 PM: npm WARN deprecated har-validator@5.1.5: this library is no longer supported
8:07:45 PM: npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
8:07:50 PM: npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
8:07:50 PM: npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
8:08:07 PM: > node-sass@4.14.1 install /opt/build/repo/node_modules/node-sass
8:08:07 PM: > node scripts/install.js
8:08:08 PM: Downloading binary from https://github.com/sass/node-sass/releases/download/v4.14.1/linux-x64-72_binding.node
8:08:08 PM: Download complete
8:08:08 PM: Binary saved to /opt/build/repo/node_modules/node-sass/vendor/linux-x64-72/binding.node
8:08:08 PM: Caching binary to /opt/buildhome/.npm/node-sass/4.14.1/linux-x64-72_binding.node
8:08:08 PM: > core-js@3.6.5 postinstall /opt/build/repo/node_modules/core-js
8:08:08 PM: > node -e "try{require('./postinstall')}catch(e){}"
8:08:09 PM: > protobufjs@6.10.2 postinstall /opt/build/repo/node_modules/protobufjs
8:08:09 PM: > node scripts/postinstall
8:08:09 PM: > node-sass@4.14.1 postinstall /opt/build/repo/node_modules/node-sass
8:08:09 PM: > node scripts/build.js
8:08:09 PM: Binary found at /opt/build/repo/node_modules/node-sass/vendor/linux-x64-72/binding.node
8:08:09 PM: Testing binary
8:08:09 PM: Binary is fine
8:08:10 PM: npm notice created a lockfile as package-lock.json. You should commit this file.
8:08:10 PM: npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.1 (node_modules/chokidar/node_modules/fsevents):
8:08:10 PM: npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
8:08:10 PM: npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/watchpack-chokidar2/node_modules/chokidar/node_modules/fsevents):
8:08:10 PM: npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
8:08:10 PM: npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/webpack-dev-server/node_modules/chokidar/node_modules/fsevents):
8:08:10 PM: npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
8:08:10 PM: npm WARN dani-bear-watcher@1.0.0 No repository field.
8:08:10 PM: added 1188 packages from 567 contributors and audited 1191 packages in 31.908s
8:08:11 PM: 67 packages are looking for funding
8:08:11 PM:   run `npm fund` for details
8:08:11 PM: found 1 high severity vulnerability
8:08:11 PM:   run `npm audit fix` to fix them, or `npm audit` for details
8:08:11 PM: NPM modules installed
8:08:11 PM: Started restoring cached go cache
8:08:11 PM: Finished restoring cached go cache
8:08:11 PM: go version go1.14.4 linux/amd64
8:08:11 PM: go version go1.14.4 linux/amd64
8:08:11 PM: Installing missing commands
8:08:11 PM: Verify run directory
8:08:13 PM: ​
8:08:13 PM: ────────────────────────────────────────────────────────────────
8:08:13 PM:   Netlify Build                                                 
8:08:13 PM: ────────────────────────────────────────────────────────────────
8:08:13 PM: ​
8:08:13 PM: ❯ Version
8:08:13 PM:   @netlify/build 9.4.0
8:08:13 PM: ​
8:08:13 PM: ❯ Flags
8:08:13 PM:   deployId: 603afae1060121b85c53686e
8:08:13 PM:   mode: buildbot
8:08:13 PM: ​
8:08:13 PM: ❯ Current directory
8:08:13 PM:   /opt/build/repo
8:08:13 PM: ​
8:08:13 PM: ❯ Config file
8:08:13 PM:   No config file was defined: using default values.
8:08:13 PM: ​
8:08:13 PM: ❯ Context
8:08:13 PM:   production
8:08:13 PM: ​
8:08:13 PM: ────────────────────────────────────────────────────────────────
8:08:13 PM:   1. Build command from Netlify app                             
8:08:13 PM: ────────────────────────────────────────────────────────────────
8:08:13 PM: ​
8:08:13 PM: $ npm run build
8:08:13 PM: > dani-bear-watcher@1.0.0 build /opt/build/repo
8:08:13 PM: > webpack --mode production --module-bind js=babel-loader
8:08:18 PM: Hash: 3b06abbc3d1d67c97b8a
8:08:18 PM: Version: webpack 4.44.1
8:08:18 PM: Time: 3805ms
8:08:18 PM: Built at: 02/28/2021 2:08:18 AM
8:08:18 PM:  17 assets
8:08:18 PM: Entrypoint main = main.css bundle.js
8:08:18 PM: [0] ./src/styles/main.scss 39 bytes {0} [built]
8:08:18 PM: [1] ./src/javascripts/main.js + 4 modules 3.8 KiB {0} [built] [1 error]
8:08:18 PM:     | ./src/javascripts/main.js 216 bytes [built]
8:08:18 PM:     | ./src/javascripts/componenets/initialScreen.js 925 bytes [built]
8:08:18 PM:     | ./src/javascripts/componenets/bearForm.js 868 bytes [built]
8:08:18 PM:     | ./src/javascripts/componenets/helpers.js 182 bytes [built]
8:08:18 PM:     | ./src/javascripts/componenets/river.js 1.66 KiB [built] [1 error]
8:08:18 PM:     + 2 hidden modules
8:08:18 PM: ERROR in ./src/javascripts/componenets/river.js
8:08:18 PM: Module Error (from ./node_modules/eslint-loader/dist/cjs.js):
8:08:18 PM: 
8:08:18 PM: /opt/build/repo/src/javascripts/componenets/river.js
8:08:18 PM:   46:65  error  'catchFish' is not defined  no-undef
8:08:18 PM: 
8:08:18 PM: ✖ 1 problem (1 error, 0 warnings)
8:08:18 PM: 
8:08:18 PM:  @ ./src/javascripts/componenets/bearForm.js 2:0-41 7:2-20
8:08:18 PM:  @ ./src/javascripts/main.js
8:08:18 PM: Child HtmlWebpackCompiler:
8:08:18 PM:      1 asset
8:08:18 PM:     Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
8:08:18 PM:     [0] ./node_modules/html-webpack-plugin/lib/loader.js!./src/index.html 256 bytes {0} [built]
8:08:18 PM: Child mini-css-extract-plugin node_modules/css-loader/dist/cjs.js??ref--7-1!node_modules/sass-loader/dist/cjs.js??ref--7-2!src/styles/main.scss:
8:08:18 PM:     Entrypoint mini-css-extract-plugin = *
8:08:18 PM:     [18] ./node_modules/css-loader/dist/cjs.js??ref--7-1!./node_modules/sass-loader/dist/cjs.js??ref--7-2!./src/styles/main.scss 430 KiB {0} [built]
8:08:18 PM:         + 18 hidden modules
8:08:18 PM: npm ERR! code ELIFECYCLE
8:08:18 PM: npm ERR! errno 2
8:08:18 PM: npm ERR! dani-bear-watcher@1.0.0 build: `webpack --mode production --module-bind js=babel-loader`
8:08:18 PM: npm ERR! Exit status 2
8:08:18 PM: npm ERR!
8:08:18 PM: npm ERR! Failed at the dani-bear-watcher@1.0.0 build script.
8:08:18 PM: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
8:08:18 PM: npm ERR! A complete log of this run can be found in:
8:08:18 PM: npm ERR!     /opt/buildhome/.npm/_logs/2021-02-28T02_08_18_719Z-debug.log
8:08:18 PM: ​
8:08:18 PM: ────────────────────────────────────────────────────────────────
8:08:18 PM:   "build.command" failed                                        
8:08:18 PM: ────────────────────────────────────────────────────────────────
8:08:18 PM: ​
8:08:18 PM:   Error message
8:08:18 PM:   Command failed with exit code 2: npm run build
8:08:18 PM: ​
8:08:18 PM:   Error location
8:08:18 PM:   In Build command from Netlify app:
8:08:18 PM:   npm run build
8:08:18 PM: ​
8:08:18 PM:   Resolved config
8:08:18 PM:   build:
8:08:18 PM:     command: npm run build
8:08:18 PM:     commandOrigin: ui
8:08:18 PM:     publish: /opt/build/repo/build
8:08:18 PM: Caching artifacts
8:08:18 PM: Started saving node modules
8:08:18 PM: Finished saving node modules
8:08:18 PM: Started saving build plugins
8:08:18 PM: Finished saving build plugins
8:08:18 PM: Started saving pip cache
8:08:18 PM: Finished saving pip cache
8:08:18 PM: Started saving emacs cask dependencies
8:08:19 PM: Finished saving emacs cask dependencies
8:08:19 PM: Started saving maven dependencies
8:08:19 PM: Finished saving maven dependencies
8:08:19 PM: Started saving boot dependencies
8:08:19 PM: Finished saving boot dependencies
8:08:19 PM: Started saving rust rustup cache
8:08:19 PM: Finished saving rust rustup cache
8:08:19 PM: Started saving go dependencies
8:08:19 PM: Finished saving go dependencies
8:08:22 PM: Build failed due to a user error: Build script returned non-zero exit code: 2
8:08:22 PM: Failing build: Failed to build site
8:08:22 PM: Failed during stage 'building site': Build script returned non-zero exit code: 2
8:08:22 PM: Finished processing build request in 51.253019124s

**Reply by danicrosby** [reply]
^^^ haha

I was able to deploy my other projects just fine, but not bear-watcher

**Reply by drteresavasquez** [reply]
There is an error in the console log you posted...it says:

opt/build/repo/src/javascripts/componenets/river.js
8:08:18 PM: 46:65 error 'catchFish' is not defined no-undef
8:08:18 PM:
8:08:18 PM: ✖ 1 problem (1 error, 0 warnings)

fix that and it should be fine.

**Reply by drteresavasquez** [reply]
Also the same error should be in your terminal when you run the app. look there first to fix it locally.

**Reply by danicrosby** [reply]
That worked! I had an unfinished stretch goal that got pushed up. I removed the half-assed stretch and re-deployed. All is well! Thanks! I didn't realize Netlify could also track this information. I have a lot to learn...

**Reply by drteresavasquez** [reply]
It literally runs what you have on the branch you selected to deploy as if you're running npm start so always make sure that your production branch is clean 😊
